ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1305,13 +1305,12 @@ private void handleGetResponse(GetResponse getResponse, List<RemoteReplicaInfo>
                     messageInfoList);
               }
 
-              remoteReplicaInfo.setToken(exchangeMetadataResponse.remoteToken);
-              remoteReplicaInfo.setLocalLagFromRemoteInBytes(exchangeMetadataResponse.localLagFromRemoteInBytes);
-              // reset stored metadata response for this replica
-              remoteReplicaInfo.setExchangeMetadataResponse(new ExchangeMetadataResponse(ServerErrorCode.No_Error));
+              if (!validMessageDetectionInputStream.hasInvalidMessages()) {
+                advanceToken(remoteReplicaInfo, exchangeMetadataResponse);
```

Comment:
What happens when we do not advance the token? I think we should either this failure visible and continue with the replication or halt the replication until we get to the bottom of the issue.

Reply:
@tbingol If we do not advance the token, then the target node (vcr) will keep fetching blobs from previous token position. And if a blob is truly corrupted, then we will never make progress be stuck in an infinite loop. Currently, there isn't a finer granularity of insight of whether a blob is truly corrupted or just the message coming from source is corrupted. Nor do we have any alerts for infinite retries. One key point of difference between server and vcr is that on server side, we emit metrics for blobs skipped due to crc error or message format exceptions. But we do not emit any metrics on vcr side, so we have no idea how many blobs we skipped. I recently added those metrics though it has not been deployed yet.